src/Features/Core/Portable/Workspace/BackgroundCompiler.cs

```diff
@@ -8,30 +8,36 @@
 using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
-using Microsoft.CodeAnalysis.Internal.Log;
-using Microsoft.CodeAnalysis.SolutionCrawler;
+using Microsoft.CodeAnalysis.Collections;
+using Microsoft.CodeAnalysis.PooledObjects;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Host
 {
     internal sealed class BackgroundCompiler : IDisposable
     {
-        private Workspace _workspace;
-        private readonly TaskQueue _taskQueue;
+        private Workspace? _workspace;
+        private readonly AsyncBatchingWorkQueue<CancellationToken> _workQueue;
```

Comment:
first: using a AsyncBatching queue to debounce things.  We use near-immediate so we can start running extremely quickly.  But this still allows us to batch up massive flurries of events without kicking off work for all of them.